src/protocol.h

```diff
@@ -354,34 +354,45 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
     return (services & NODE_NETWORK) || (services & NODE_NETWORK_LIMITED);
 }
 
+enum class AddrSerialization {
+    DISK,
+    NETWORK_NOTIME,
+    NETWORK_WITHTIME,
+};
+
 /** A CService with information about it as peer */
 class CAddress : public CService
 {
     static constexpr uint32_t TIME_INIT{100000000};
 
+    /** The disk serialization for CAddress includes a version number.
+     *  Traditionally the number CLIENT_VERSION was used, but it has since
+     *  been disentangled from it. */
+    static constexpr uint32_t DISK_VERSION{210000};
+
 public:
     CAddress() : CService{} {};
     explicit CAddress(CService ipIn, ServiceFlags nServicesIn) : CService{ipIn}, nServices{nServicesIn} {};
 
-    SERIALIZE_METHODS(CAddress, obj)
+    SERIALIZE_METHODS_PARAMS(CAddress, obj, AddrSerialization, fmt)
     {
         SER_READ(obj, obj.nTime = TIME_INIT);
```

Comment:
This has been added here on purpose: https://github.com/bitcoin/bitcoin/pull/19020#discussion_r427620111

Reply:
The linked comment doesn't make much sense to me. There are plenty of examples where member fields aren't [de]serialized, eg:

https://github.com/bitcoin/bitcoin/blob/0f6900e780afb9c176cf7d1da4a1fc0ca6c9118a/src/addrman.h#L48-L55
https://github.com/bitcoin/bitcoin/blob/0f6900e780afb9c176cf7d1da4a1fc0ca6c9118a/src/primitives/block.h#L68-L69
https://github.com/bitcoin/bitcoin/blob/0f6900e780afb9c176cf7d1da4a1fc0ca6c9118a/src/wallet/walletutil.h#L99

or only [de]serialized conditionally:

https://github.com/bitcoin/bitcoin/blob/0f6900e780afb9c176cf7d1da4a1fc0ca6c9118a/src/wallet/walletdb.h#L106-L108

Reply:
Could it happen that we unserialize into the same object twice? E.g.
```
s >> addr;
... do something with addr, possibly change nTime ...
s >> addr;
// if we don't explicitly set nTime during unserialization,
// then here we may end up with some leftover nTime?
```

Reply:
Yes, that's the reason to prefer fully initializing all fields when deserializing (or at least, all fields that are subject to serialization in the first place).

Reply:
@jnewbery I think in general we should aim to support deserializing into an already existing (and not just default-initialized) objects, as it's hard to enforce the opposite.

`CAddrInfo` is a bit of a special case, as its memory-only fields are "managed" by CAddrMan more than being a standalone object. In the other cases, I think it makes sense to make deserialization reset those fields.

Reply:
Thanks for the additional background. My immediate reaction is that it seems dangerous to re-deserialize into an existing object and that it could lead to subtle, confusing bugs. However, we've definitely wondered off-topic for this PR, so I'm going to mark this conversation as resolved. If the expectation really is that the deserialization methods overwrite all fields, a future PR could update the documentation for `SERIALIZE_METHODS` to explicitly state that.